components/core/src/clp/StreamingReader.cpp

```diff
@@ -0,0 +1,321 @@
+}
+}  // namespace
+
+bool StreamingReader::m_initialized{false};
+
+auto StreamingReader::init() -> ErrorCode {
+    if (m_initialized) {
+        return ErrorCode_Success;
+    }
+    if (0 != curl_global_init(CURL_GLOBAL_ALL)) {
+        return ErrorCode_Failure;
+    }
+    m_initialized = true;
+    return ErrorCode_Success;
+}
+
+auto StreamingReader::deinit() -> void {
+    curl_global_cleanup();
+    m_initialized = false;
+}
+
+auto StreamingReader::TransferThread::thread_method() -> void {
+    try {
+        CurlDownloadHandler curl_handler{
+                m_reader.m_src_url,
+                static_cast<void*>(&m_reader),
+                curl_progress_callback,
+                curl_write_callback,
+                static_cast<long>(m_reader.m_connection_timeout),
+                static_cast<long>(m_reader.m_overall_timeout),
+                m_offset,
+                m_disable_caching
+        };
+        m_reader.m_curl_return_code = curl_handler.perform();
+        m_reader.enqueue_filled_buffer();
+        m_reader.set_state_code(
+                (CURLE_OK == m_reader.m_curl_return_code) ? State::Finished : State::Failed
+        );
+    } catch (CurlOperationFailed const& ex) {
+        m_reader.m_curl_return_code = ex.get_curl_err();
+        m_reader.set_state_code(State::Failed);
+    }
+
+    m_reader.m_cv_reader.notify_all();
+}
+
+StreamingReader::StreamingReader(
+        std::string_view src_url,
+        size_t offset,
+        bool disable_caching,
+        uint32_t overall_timeout_int_sec,
+        uint32_t connection_timeout_in_sec,
+        size_t buffer_pool_size,
+        size_t buffer_size
+)
+        : m_overall_timeout{overall_timeout_int_sec},
+          m_connection_timeout{connection_timeout_in_sec},
+          m_buffer_pool_size{std::max(cMinBufferPoolSize, buffer_pool_size)},
+          m_buffer_size{std::max(cMinBufferPoolSize, buffer_size)} {
+    for (size_t i = 0; i < m_buffer_pool_size; ++i) {
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+        m_buffer_pool.emplace_back(std::make_unique<char[]>(m_buffer_size));
+    }
+    if (false == m_initialized) {
+        throw OperationFailed(ErrorCode_NotReady, __FILE__, __LINE__);
+    }
+    m_src_url = src_url;
+    m_transfer_thread = std::make_unique<TransferThread>(std::ref(*this), offset, disable_caching);
+    m_transfer_thread->start();
+}
+
+StreamingReader::~StreamingReader() {
+    if (is_curl_transfer_in_progress()) {
+        // We need to kill the current connected session to reclaim CURL resources. Since the
+        // transfer thread is asynchronized, we need to abort the session to stop data transfer.
+        abort_data_transfer();
+    }
+
+    while (is_transfer_thread_running()) {
+        // TODO: we could use sleep instead of pulling
+    }
+}
+
+auto StreamingReader::abort_data_transfer() -> void {
+    m_transfer_aborted.store(true);
+    m_cv_transfer.notify_all();
```

Comment:
Sorry, forgot to mention this in the previous review:

I think there's a race condition if we don't call this within the monitor lock, right?

For instance:

* There are no empty buffers.
* The background thread calls the progress callback and write callback but before it goes to sleep, there's a context switch to the main thread.
* The main thread calls `abort_data_transfer` which sends a notification to all waiting threads.
* There's a context switch to the background thread and it goes to sleep since it missed the notification.

Reply:
right, they should be locked. My bad